crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs

```diff
@@ -701,3 +707,233 @@ pub(crate) fn numpy_2_0_deprecation(checker: &mut Checker, expr: &Expr) {
     };
     checker.diagnostics.push(diagnostic);
 }
+
+/// Ignore attempts to access a `numpy` member via its deprecated name
+/// if the access takes place in an `except` block that provides compatibility
+/// with older numpy versions.
+///
+/// For attribute accesses (e.g. `np.ComplexWarning`), we only ignore the violation
+/// if it's inside an `except AttributeError` block, and the member is accessed
+/// through its non-deprecated name in the associated `try` block.
+///
+/// For uses of the `numpy` member where it's simply an `ExprName` node,
+/// we check to see how the `numpy` member was bound. If it was bound via a
+/// `from numpy import foo` statement, we check to see if that import statement
+/// took place inside an `except ImportError` or `except ModuleNotFoundError` block.
+/// If so, and if the `numpy` member was imported through its non-deprecated name
+/// in the associated try block, we ignore the violation in the same way.
+///
+/// Examples:
+///
+/// ```py
+/// import numpy as np
+///
+/// try:
+///     np.all([True, True])
+/// except AttributeError:
+///     np.alltrue([True, True])  # Okay
+///
+/// try:
+///     from numpy.exceptions import ComplexWarning
+/// except ImportError:
+///     from numpy import ComplexWarning
+///
+/// x = ComplexWarning()  # Okay
+/// ```
+fn is_guarded_by_try_except(
+    expr: &Expr,
+    replacement: &Replacement,
+    semantic: &SemanticModel,
+) -> bool {
+    match expr {
+        Expr::Attribute(_) => {
+            if !semantic.in_exception_handler() {
+                return false;
+            }
+            let Some(try_node) = semantic
+                .current_statements()
+                .find_map(|stmt| stmt.as_try_stmt())
+            else {
+                return false;
+            };
+            let suspended_exceptions = Exceptions::from_try_stmt(try_node, semantic);
+            if !suspended_exceptions.contains(Exceptions::ATTRIBUTE_ERROR) {
```

Comment:
Can we check `semantic.handled_exceptions` instead? Does that work? Then we wouldn't need to find the try statement or anything.

Reply:
No, `semantic.handled_exceptions` records the exceptions if we're visiting the `try` block of a `try`/`except` statement. But here we're visiting the `except` block of a `try`/`except` statement. Hence "suspended_exceptions" rather than "handled_exceptions".

Reply:
Also we still need to find the node for the `try` statement so that we can check whether it was accessed via the undeprecated call path in the `try` block